File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys
n, m = map(int, input().split())

check = [False] * (n+1)
ans = [0] * m

# index번째의 수열 결정하는 함수
def P(n, m, index):
    if index == m:
        sys.stdout.write(' '.join(map(str, ans)) + '\n')
        return
    for i in range(1, n+1):
        if check[i]:
            continue
        check[i] = True
        ans[index] = i
        P(n, m, index+1)
        check[i] = False

P(n, m, 0)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.js
const fs = require("fs");

const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";

const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [N, M] = input[0].split(" ").map((v) => Number(v));
const answer = Array(10).fill(0);
const visited = Array(10).fill(false);

function dfs(k) {
  if (k == M) {
    console.log(answer.filter((_, i) => i < M).join(" "));
    return;
  }

  for (let i = 1; i <= N; i++) {
    if (!visited[i]) {
      visited[i] = true;
      answer[k] = i;
      dfs(k + 1);
      visited[i] = false;
    }
  }
}

dfs(0);
